domain/usecases/subscribe_test.go

```diff
@@ -0,0 +1,146 @@
+package usecases
+
+import (
+	"context"
+	"testing"
+
+	"github.com/matheusmosca/walrus/domain/entities"
+	"github.com/matheusmosca/walrus/domain/vos"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSubsribe(t *testing.T) {
+	t.Parallel()
+
+	type args struct {
+		ctx       context.Context
+		topicName vos.TopicName
+	}
+
+	type fields struct {
+		storage   Repository
+		topicName vos.TopicName
+	}
+
+	tests := []struct {
+		name    string
+		args    args
+		fields  fields
+		wantCh  chan vos.Message
+		wantId  vos.SubscriberID
+		wantErr error
+	}{
+		{
+			name: "subscribe should succeed",
+			args: args{
+				ctx:       context.Background(),
+				topicName: "walrus",
+			},
+			fields: fields{
+				storage:   &RepositoryMock{},
+				topicName: "walrus",
+			},
+			wantCh:  make(chan vos.Message),
+			wantId:  vos.SubscriberID("id"),
+			wantErr: nil,
+		},
+		{
+			name: "short topic name should return error",
+			args: args{
+				ctx:       context.Background(),
+				topicName: "xd",
+			},
+			fields: fields{
+				storage:   &RepositoryMock{},
+				topicName: "walrus",
+			},
+			wantCh:  nil,
+			wantId:  "",
+			wantErr: vos.ErrTopicNameTooShort,
+		},
+		{
+			name: "empty topic name message should return error",
+			args: args{
+				ctx:       context.Background(),
+				topicName: "",
+			},
+			fields: fields{
+				storage:   &RepositoryMock{},
+				topicName: "walrus",
+			},
+			wantCh:  nil,
+			wantId:  "",
+			wantErr: vos.ErrEmptyTopicName,
+		},
+		{
+			name: "nonexistent topic should return error",
+			args: args{
+				ctx:       context.Background(),
+				topicName: "testtopic",
+			},
+			fields: fields{
+				storage:   &RepositoryMock{},
+				topicName: "walrus",
+			},
+			wantCh:  nil,
+			wantId:  "",
+			wantErr: entities.ErrTopicNotFound,
+		},
+		{
+			name: "topic already exists should return error",
+			args: args{
+				ctx:       context.Background(),
+				topicName: "walrus",
+			},
+			fields: fields{
+				storage:   &RepositoryMock{},
+				topicName: "walrus",
+			},
+			wantCh:  nil,
+			wantId:  "",
+			wantErr: entities.ErrTopicNotFound,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			// init the topic
+			topic, err := entities.NewTopic(tt.fields.topicName)
+			require.NoError(t, err)
+
+			// activate the topic
+			topic.Activate()
+
+			tt.fields.storage = &RepositoryMock{
+				GetTopicFunc: func(ctx context.Context, topicName vos.TopicName) (entities.Topic, error) {
+					if tt.fields.topicName != tt.args.topicName {
+						return entities.Topic{}, entities.ErrTopicNotFound
+					}
+
+					return topic, nil
+				},
+				CreateTopicFunc: func(ctx context.Context, topicName vos.TopicName, topic entities.Topic) error {
+					if tt.fields.topicName == tt.args.topicName {
+						return entities.ErrTopicAlreadyExists
+					}
+
+					return nil
+				},
+			}
```

Comment:
Could you define these mocks in each test scenario present in the slice of tests? You could change the `fields` to be a function that returns a `RepositoryMock` and a `entities.Topic` like this:
```go
// before
type fields struct {
  storage      Repository
  topicName vos.TopicName
}
// after
func (t *testing.T) (Repository, *entities.Topic) {}
```

Reply:
Hi @matheusmosca, I couldn't make updates exactly like you suggested because I was getting this "mixture of field:value and value initializers" error. Also, my UTs for the success cases were getting skipped during local testing too and I'm not sure why. Can you take a look?

Reply:
@quanuw you can start from the example below by adding failures scenarios. I've noticed some race conditions issues so do not change or remove the time.Sleep(), I'll investigate it further later. But for now, it does the job.

```go
package usecases

import (
	"context"
	"testing"
	"time"

	"github.com/matheusmosca/walrus/domain/entities"
	"github.com/matheusmosca/walrus/domain/vos"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestSubscribe(t *testing.T) {
	t.Parallel()
	type args struct {
		ctx       context.Context
		topicName vos.TopicName
		message   vos.Message
	}
	tests := []struct {
		name        string
		args        args
		fields      func(t *testing.T) (Repository, entities.Topic)
		wantMessage vos.Message
		wantErr     error
	}{
		{
			name: "subscribe should succeed",
			args: args{
				ctx:       context.Background(),
				topicName: "walrus",
				message: vos.Message{
					TopicName:   vos.TopicName("walrus"),
					PublishedBy: "test_publisher",
					Body:        []byte("hello world"),
				},
			},
			fields: func(t *testing.T) (Repository, entities.Topic) {
				topic, err := entities.NewTopic("walrus")
				require.NoError(t, err)

				repoMock := &RepositoryMock{
					GetTopicFunc: func(ctx context.Context, topicName vos.TopicName) (entities.Topic, error) {
						return topic, nil
					},
				}

				return repoMock, topic
			},
			wantMessage: vos.Message{
				TopicName:   vos.TopicName("walrus"),
				PublishedBy: "test_publisher",
				Body:        []byte("hello world"),
			},
			wantErr: nil,
		},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			repo, topic := tt.fields(t)
			topic.Activate()

			useCase := New(repo)
			subCh, id, err := useCase.Subscribe(tt.args.ctx, tt.args.topicName)
			assert.ErrorIs(t, err, tt.wantErr)

			if tt.wantErr != nil {
				assert.Empty(t, subCh)
				assert.Empty(t, id)
				return
			}
			defer close(subCh)

			assert.NotEmpty(t, id)

			go func() {
				time.Sleep(time.Millisecond * 4)
				err = topic.Dispatch(tt.args.message)
				require.NoError(t, err)
			}()

			gotMessage := <-subCh
			assert.Equal(t, tt.wantMessage, gotMessage)
		})
	}
}

```

Reply:
Thanks for looking! I updated the code, but the "nonexistent topic subscribe should succeed" is still hitting a deadlock. Any ideas here?